openmc/deplete/results.py

```diff
@@ -0,0 +1,496 @@
+"""
+
+from collections import OrderedDict
+import copy
+from warnings import warn
+
+import numpy as np
+import h5py
+
+from . import comm, have_mpi, MPI
+from .reaction_rates import ReactionRates
+
+_VERSION_RESULTS = (1, 0)
+
+
+__all__ = ["Results"]
+
+
+class Results(object):
+    """Output of a depletion run
+
+    Attributes
+    ----------
+    k : list of (float, float)
+        Eigenvalue and uncertainty for each substep.
+    time : list of float
+        Time at beginning, end of step, in seconds.
+    power : float
+        Power during time step, in Watts
+    n_mat : int
+        Number of mats.
+    n_nuc : int
+        Number of nuclides.
+    rates : list of ReactionRates
+        The reaction rates for each substep.
+    volume : OrderedDict of int to float
+        Dictionary mapping mat id to volume.
+    mat_to_ind : OrderedDict of str to int
+        A dictionary mapping mat ID as string to index.
+    nuc_to_ind : OrderedDict of str to int
+        A dictionary mapping nuclide name as string to index.
+    mat_to_hdf5_ind : OrderedDict of str to int
+        A dictionary mapping mat ID as string to global index.
+    n_hdf5_mats : int
+        Number of materials in entire geometry.
+    n_stages : int
+        Number of stages in simulation.
+    data : numpy.ndarray
+        Atom quantity, stored by stage, mat, then by nuclide.
+    proc_time: int
+        Average time spent depleting a material across all
+        materials and processes
+
+    """
+    def __init__(self):
+        self.k = None
+        self.time = None
+        self.power = None
+        self.rates = None
+        self.volume = None
+        self.proc_time = None
+
+        self.mat_to_ind = None
+        self.nuc_to_ind = None
+        self.mat_to_hdf5_ind = None
+
+        self.data = None
+
+    def __getitem__(self, pos):
+        """Retrieves an item from results.
+
+        Parameters
+        ----------
+        pos : tuple
+            A three-length tuple containing a stage index, mat index and a nuc
+            index.  All can be integers or slices.  The second two can be
+            strings corresponding to their respective dictionary.
+
+        Returns
+        -------
+        float
+            The atoms for stage, mat, nuc
+
+        """
+        stage, mat, nuc = pos
+        if isinstance(mat, str):
+            mat = self.mat_to_ind[mat]
+        if isinstance(nuc, str):
+            nuc = self.nuc_to_ind[nuc]
+
+        return self.data[stage, mat, nuc]
+
+    def __setitem__(self, pos, val):
+        """Sets an item from results.
+
+        Parameters
+        ----------
+        pos : tuple
+            A three-length tuple containing a stage index, mat index and a nuc
+            index.  All can be integers or slices.  The second two can be
+            strings corresponding to their respective dictionary.
+
+        val : float
+            The value to set data to.
+
+        """
```

Comment:
We should check the type here with `cv.check_type` since, without it, we could just add anything to our data.